pkg/reconciler/service/resources/rolloutorchestrator.go

```diff
@@ -0,0 +1,202 @@
+type RevisionRecord struct {
+	MinScale *int32
+	MaxScale *int32
+	Name     string
+}
+
+// ReadIntAnnotation reads the int value of a specific key in the annotation of the revision.
+func ReadIntAnnotation(revision *servingv1.Revision, key string) (result *int32) {
+	if val, ok := revision.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntServiceAnnotation reads the int value of a specific key in the annotation of the service.
+func ReadIntServiceAnnotation(service *servingv1.Service, key string) (result *int32) {
+	if val, ok := service.Spec.Template.ObjectMeta.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntRevisionRecord reads the minScale and maxScale in the RevisionRecord.
+func ReadIntRevisionRecord(val RevisionRecord) (min *int32, max *int32) {
+	if val.MinScale != nil {
+		min = ptr.Int32(*val.MinScale)
+	}
+	if val.MaxScale != nil {
+		max = ptr.Int32(*val.MaxScale)
+	}
+	return min, max
+}
+
+func initializeTargetRevisions(ultimateRevisionTarget *[]v1.TargetRevision, traffic *servingv1.TrafficTarget,
+	index int, lastRevName string, service *servingv1.Service, records map[string]RevisionRecord) {
+	target := v1.TargetRevision{}
+	if traffic.RevisionName == "" || traffic.RevisionName == lastRevName {
+		target.LatestRevision = ptr.Bool(true)
+		target.RevisionName = lastRevName
+	} else {
+		target.LatestRevision = ptr.Bool(false)
+		target.RevisionName = traffic.RevisionName
+	}
+	if traffic.Percent == nil {
+		target.Percent = ptr.Int64(100)
+	} else {
+		target.Percent = ptr.Int64(*traffic.Percent)
+	}
+
+	target.ConfigurationName = traffic.ConfigurationName
+	target.Tag = traffic.Tag
+	target.URL = traffic.URL
+	if val, ok := records[target.RevisionName]; ok {
+		target.MinScale, target.MaxScale = ReadIntRevisionRecord(val)
+	} else {
+		// Get min and max scales from the service
+		target.MinScale = ReadIntServiceAnnotation(service, autoscaling.MinScaleAnnotationKey)
+		target.MaxScale = ReadIntServiceAnnotation(service, autoscaling.MaxScaleAnnotationKey)
+	}
+	(*ultimateRevisionTarget)[index] = target
+}
+
+// GetInitialFinalTargetRevision is used to generate the initialTargetRevision and ultimateRevisionTarget.
+// Both of them are needed for the RolloutOrchestrator creation.
+// Only ultimateRevisionTarget for the RolloutOrchestrator update.
+func GetInitialFinalTargetRevision(service *servingv1.Service, records map[string]RevisionRecord,
+	route *servingv1.Route) ([]v1.TargetRevision, []v1.TargetRevision) {
+	var initialTargetRevision, ultimateRevisionTarget []v1.TargetRevision
+	// This is how the last revision is named after the service generation.
+	lastRevName := kmeta.ChildName(service.Name, fmt.Sprintf("-%05d", service.Generation))
```

Comment:
The last created revision name should be also stored in service status

Reply:
Not really, because at the very beginning, I create the RO even before the configuration's reconciliation. The RO creation is at the beginning.